### PR TITLE
Updated Implementation/Services + /Implementation/Composing to v9

### PR DESCRIPTION
**Pull request by bergmania**
Updated Implementation/Services + /Implementation/Composing to v9.

Mainly code examples that had to change

### CODE REVIEW

File: Implementation/Composing/index-v9.md [hunk 1]
```diff
@@ -0,0 +1,519 @@
...
+            // Change the OEmbedProviders collection
+            // by adding our new EmbedProvider for Spotify
+            builder.OEmbedProviders().Append<Spotify>();
+        }
+    }
+}
+```
+See a list of collections below to determine which are 'type scanned' and which require explicit registration.
+
+### Example - Creating a Composer to listen for ContentSavingNotification
```
Comment (p-m-j) [inline Implementation/Composing/index-v9.md:99]: I think this example should go above the spotify stuff, far more relatable and common use case

File: Implementation/Composing/index-v9.md [hunk 2]
```diff
@@ -0,0 +1,519 @@
...
+    }
+}
+```
+
+
+## Attributes
+Umbraco has some useful C# attributes to decorate your composer classes or Types used in collections, to give you further control on how and when your Composers will 'compose'.
+
+### `[ComposeBefore]` and `[ComposeAfter]`
+A finer-grain mechanism can then be used to refine the order of composition. Each composer can specify that it should compose before or after another composer, using the ComposeBefore and ComposeAfter attributes. For instance:
```
Comment (p-m-j) [inline Implementation/Composing/index-v9.md:228]: Hopefully we can nuke this section, perhaps with a section to talk about alternative approaches

File: Implementation/Composing/index-v9.md [hunk 3]
```diff
@@ -0,0 +1,519 @@
...
+    }
+}
+```
+
+### [HideFromTypeFinder]
+
+This is used to hide a type from being auto-scanned/added to a collection as in some cases certain items/types may need to be added to a collection manually. For example, a Search package may make it optional whether to replace the 'backoffice search' with an ISearchableTree implementation. Type scanning would make this change automatically at start up if the custom implementation was detected via type scanning. This attribute could hide the class from the scanner.
+
+### [DisableComposer] & [Disable]
+
```
Comment (p-m-j) [inline Implementation/Composing/index-v9.md:273]: Hopefully we can nuke this section, perhaps with a section to talk about alternative approaches

File: Implementation/Composing/index-v9.md (hunk 1 above)
Comment (bergmania) [reply Implementation/Composing/index-v9.md:99]: That would be okay with me, but I kept the order from the v8 document.. Maybe that change should also be reflected into the v8 docs

Comment (p-m-j) [reply Implementation/Composing/index-v9.md:99]: Yep think point stands for both versions :)

File: Implementation/Composing/index-v9.md [hunk 4]
```diff
@@ -0,0 +1,519 @@
+---
+keywords: composing composers components runtime boot booting v8 version8 events registering
```
Comment (jonat123) [inline Implementation/Composing/index-v9.md:2]:
```suggestion
keywords: composing composers components runtime boot booting v9 version9 events registering
```

File: Implementation/Composing/index-v9.md [hunk 5]
```diff
@@ -0,0 +1,519 @@
...
+            }
+        }
+    }
+}
+```
+
+:::warning
+Ordering of composers is important, the last one added can override a previously added composer! Make sure, when overriding,
+that your composer that is doing the overriding, is 'composing', after the composer has 'composed' the element you wish to override!
+:::
```
Comment (jonat123) [inline Implementation/Composing/index-v9.md:142]:
```suggestion
### Example - Creating a Composer to listen for ContentSavingNotification

This example shows how to create a component and an notification handler for the `ContentSavingNotification`, (perhaps to check for explicit words, or some custom business logic that needs to run before the content item is saved in Umbraco).

We create a new C# class that implements `IComposer` and use it register our notification handler.


```csharp
using System.Linq;
using Umbraco.Cms.Core.Composing;
using Umbraco.Cms.Core.DependencyInjection;
using Umbraco.Cms.Core.Events;
using Umbraco.Cms.Core.Notifications;
using Umbraco.Extensions;

namespace My.Website
{
    public class SubscribeToContentServiceSavingComposer : IComposer
    {
        public void Compose(IUmbracoBuilder builder)
        {
            builder.AddNotificationHandler<ContentSavingNotification, CustomContentSavingNotificationHandler>()
        }
    }

    public class CustomContentSavingNotificationHandler : INotificationHandler<ContentSavingNotification>
    {
        public void Handle(ContentSavingNotification notification)
        {
            foreach (var content in notification.SavedEntities
                // Check if the content item type has a specific alias
                .Where(c => c.ContentType.Alias.InvariantEquals("MyContentType")))
            {
                // Do something if the content is using the MyContentType doctype
            }
        }
    }
}
```

:::warning
Ordering of composers is important, the last one added can override a previously added composer! Make sure, when overriding,
that your composer that is doing the overriding, is 'composing', after the composer has 'composed' the element you wish to override!
:::

### Example - Explicitly Registering a new custom OEmbedProvider
This example shows a custom 'Spotify' OEmbed Provider which will allow Spotify URLs to be used via the 'embed' button in the Grid and Rich Text Editors. As the collection for OEmbedProviders is not 'typed scanned', we need to explicitly register the provider in the collection of OEmbedProviders. We create a C# class which implements `IUserComposer` and append our new Spotify OEmbedProvider to the OEmbedProviders() collection:

```csharp

using System.Collections.Generic;
using Umbraco.Cms.Core.Media.EmbedProviders;
using Umbraco.Cms.Core.Serialization;

namespace My.Website
{
    public class Spotify : EmbedProviderBase
    {
        public Spotify(IJsonSerializer jsonSerializer)
            : base(jsonSerializer)
        {
        }

        public override string ApiEndpoint => "https://embed.spotify.com/oembed/";

        // Playlist
        // https://open.spotify.com/user/spotify/playlist/37i9dQZF1E4sNI4jZloSZr?si=cueBooBfTnqCGriSa4N_Kg
        // spotify:user:spotify:playlist:37i9dQZF1E4sNI4jZloSZr

        // Artist
        // https://open.spotify.com/artist/0iirUbtgwt9jEkc2Grin8C?si=TLeUR2cHR-KPRJJhW6YiVg
        // spotify:artist:0iirUbtgwt9jEkc2Grin8C

        // Album
        // https://open.spotify.com/album/0lvtdqkqIln6uDBBUT7DHL?si=XTVJIEmnS_OVv9l6ktPFiw
        // spotify:album:0lvtdqkqIln6uDBBUT7DHL

        // Track
        // https://open.spotify.com/track/7aCk4XfXIEJM2MecU6Gmf2?si=vESDzI0xTNeA9FQ_dvf1eQ
        // spotify:track:7aCk4XfXIEJM2MecU6Gmf2

        public override string[] UrlSchemeRegex => new string[]
        {
            @".*.spotify.com/.*",
            @"spotify:.*"
        };

        public override Dictionary<string, string> RequestParams => new Dictionary<string, string>();

        public override string GetMarkup(string url, int maxWidth = 0, int maxHeight = 0)
        {
            var requestUrl = base.GetEmbedProviderUrl(url, maxWidth, maxHeight);
            var oembed = base.GetJsonResponse<OEmbedResponse>(requestUrl);

            return oembed.GetHtml();
        }
    }
}
```

```csharp
using Umbraco.Cms.Core.Composing;
using Umbraco.Cms.Core.DependencyInjection;

namespace My.Website
{
    public class CustomOEmbedComposer : IComposer
    {
        public void Compose(IUmbracoBuilder builder)
        {
            // Change the OEmbedProviders collection
            // by adding our new EmbedProvider for Spotify
            builder.OEmbedProviders().Append<Spotify>();
        }
    }
}
```
See a list of collections below to determine which are 'type scanned' and which require explicit registration.
```

**Comment by jonat123** [comment]
@bergmania Looks good to me, I have made some small corrections for the keywords as they mentioned V8 and not 9 and moved the Spotify example down so that the one after is first, i have committed the changes and I'll merge this into our documentation 🙂

**Pull request merged**